### DISCUSSION TITLE
What's the recommended way to use SmartFormat in Unity without the Unity Localization package?

### DISCUSSION THREAD

**Discussion by jakeboxer** [axuno/SmartFormat#348] [Q&A] [answered]
We recently switched from using the Unity Localization package to our own custom-built localization solution. However, we really liked SmartFormat, and we'd like to keep using it.

Right now, we're just keeping the Unity Localization package installed purely for SmartFormat, but I'd love to remove it and add a separate SmartFormat package. Other C# libraries achieve this by creating a UPM package (example: [ink-Unity integration](https://openupm.com/packages/com.inkle.ink-unity-integration/)), but it looks like this isn't available for SmartFormat (aside from [this one](https://openupm.com/packages/trismegistus.unity.smartformat/), which is unofficial and very out of date).

Is there an alternative method you'd recommend for pulling SmartFormat into a Unity project? If not, would you consider creating an official SmartFormat UPM package?

**Comment by axunonb** [comment]
Hi, and thanks for falling in love with `SmartFormat`. Yes, I'm aware that `SmartFormat` is used in the Unity Localization package. Our latest major version was inspired by some modifications that were introduced there.
Unfortunately I'm not working with Unity, but still supplying an UPM package sounds nice. [This extension](https://openupm.com/packages/trismegistus.unity.smartformat/) is already 4 years old. Still it might be a good starting point. Of course it would be great, if new UPM versions were produced automatically in the CI workflow.
@karljj1 What are your thoughts on Unity integration (either way)?

**Reply by karljj1** [reply, marked as answer]
The localization package version does have several changes made to support Unity(serialization) and the .Net version. I do have plans to try and integrate it closer into the Unity editor, maybe as a module however there's nothing concrete at the moment. You could just drag the Smart format folder out of the localization package, you may need to make a few small changes to remove some references to the localization package.
Why did you decide to switch to your own localization solution?

**Reply by jakeboxer** [reply]
> I do have plans to try and integrate it closer into the Unity editor, maybe as a module however there's nothing concrete at the moment. You could just drag the Smart format folder out of the localization package, you may need to make a few small changes to remove some references to the localization package.

Good to know, I'll keep an eye out for the module or UPM package! Until then, I think I'll stick with the version built into the Unity Localization package so we have an easier time getting updates and stuff.

> Why did you decide to switch to your own localization solution?

The Unity Localization package is very full-featured, but a lot of the features added friction to our workflow. The main culprits were working in the Unity Localization Tables editor (some actions in it are a bit clunky) and reading Git diffs.

I think those points of friction would be acceptable for different team makeups, but only 5 members of our 9-person company need to touch the localization strings, and all 5 of us know how to program. With that makeup, a few simple JSON files are a lot lower-friction for us to work with.

**Comment by EpsilonD3lta** [comment]
This would be nice. However v3 uses some features only available in c# 9 and c# 10, which are not currently supported in Unity. Even version 2.7.2 generates warnings because Unity does not support nullable reference out of box. And it generates 1 error due to JsonSource which Unity does not recognize

**Comment by axunonb** [comment]
> v3 uses some features only available in c# 9 and c# 10

After compiling the current SmartFormat core release with the C# 8 compiler, it looks like the features causing compiler errors are:
* File scoped namespace - we could consider switching back to block scope
* Nullable type parameter in class `Variable<T>` - supported with C#9

Since v2021.2 Unity is using C# 9 isn't it. Which Unity version are you aiming at?

**Comment by EpsilonD3lta** [comment]
I am using Unity 2022.3 and you are right after closer look, the 171 errors about unsupported features are all the same - unsupported file scoped namespaces. 
Just for the sake of completeness, here are the features from c# 9 that Unity still does not support:
https://docs.unity3d.com/Manual/CSharpCompiler.html
But as you wrote, none of them are used.

However: 
- v3 also seems to use Unsafe code, which is also not supported out box, but can be toggled in Project Settings
- Unity does not support System.Text.Json, so that must be removed
- Many test scripts do not compile. I did not investigate why (missing nugets?)

So I think it's doable with some little work. Maybe a Unity branch or fork?

**Comment by axunonb** [comment]
> * v3 also seems to use Unsafe code, which is also not supported out box, but can be toggled in Project Settings

That's correct and is caused by `ZString`

> * Unity does not support System.Text.Json, so that must be removed

Simply install the the **core NuGet package** https://www.nuget.org/packages/SmartFormat/

> * Many test scripts do not compile. I did not investigate why (missing nugets?)

That's too general to comment.

**Comment by EpsilonD3lta** [comment]
> Simply install the the core NuGet package https://www.nuget.org/packages/SmartFormat/

Unity does not support NuGet packages, and if it did, how would that help with that problem?

Btw, I am commenting here because I like the idea of SmartFormat being upm package or at least Unity-ready (and of course I like SmartFormat), not because I need help to port it myself. I don't use it at the moment, but tried to port it and it's fairly easy. 
Just listing all the things that would have to be done to make it Unity-ready as the author of this issue wanted
